app/controllers/KairosDBProxy.java

```diff
@@ -0,0 +1,349 @@
+                    .thenApply(KairosMetricNamesQueryResponse::getResults)
+                    .thenApply(list -> {
+                        _cache.put(METRICS_KEY, list);
+                        _metricsList.set(list);
+                        return list;
+                    });
+
+            final List<String> metricsList = _metricsList.get();
+            if (metricsList != null) {
+                response = CompletableFuture.completedFuture(metricsList);
+            } else {
+                response = queryResponse;
+            }
+        }
+
+        return response
+                .thenApply(list -> filterMetricNames(list, containing))
+                .thenApply(list -> new KairosMetricNamesQueryResponse.Builder().setResults(list).build())
+                .<JsonNode>thenApply(_mapper::valueToTree)
+                .thenApply(Results::ok);
+    }
+
+    private static List<String> filterMetricNames(final List<String> metricNames, final String containing) {
+        if (containing == null || containing.isEmpty()) {
+            return metricNames;
+        }
+
+        final String lowerContaining = containing.toLowerCase(Locale.ENGLISH);
+        final Predicate<String> containsPredicate = s -> s.toLowerCase().contains(lowerContaining);
+        final Predicate<String> isPT1M = s -> s.startsWith("PT1M/");
+
+        return metricNames.stream()
+                .filter(isPT1M.negate())
+                .filter(containsPredicate)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Proxy a request.
+     *
+     * @return the proxied {@link Result}
+     */
+    private CompletionStage<Result> proxy() {
+        final String path = request().uri();
+        LOGGER.info().setMessage("proxying call from " + path + " to kairosdb").log();
+        final CompletableFuture<Result> promise = new CompletableFuture<>();
+        final Http.Request request = request();
+        final boolean isHttp10 = request.version().equals("HTTP/1.0");
+        LOGGER.info().setMessage(String.format("Version=%s", request.version())).log();
+        try {
+            final PipedOutputStream outputStream = new PipedOutputStream();
+            final PipedInputStream inputStream = new PipedInputStream(outputStream);
+            final Http.Response configResponse = response();
+            _client.proxy(
+                    path.startsWith("/") ? path : "/" + path,
+                    request,
+                    new ResponseHandler(outputStream, configResponse, inputStream, promise, isHttp10));
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+        return promise;
+    }
+
+    private final ProxyClient _client;
+    private final KairosDbClient _kairosDbClient;
+    private final ObjectMapper _mapper;
+    private final Cache<String, List<String>> _cache = CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.MINUTES).build();
```

Comment:
The cache has a built-in rebuild handler. Did not that work?

Reply:
The threading model is not what we want for that to work properly.  Using the built-in reloader would block the thread during a request.